flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/file/ProducerMergePartitionFileWriter.java

```diff
@@ -0,0 +1,182 @@
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.apache.flink.runtime.io.network.partition.hybrid.tiered.common.TieredStorageUtils.generateBufferWithHeaders;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/** THe implementation of {@link PartitionFileWriter} with merged logic. */
+public class ProducerMergePartitionFileWriter implements PartitionFileWriter {
+
+    private static final Logger LOG =
+            LoggerFactory.getLogger(ProducerMergePartitionFileWriter.class);
+
+    /** One thread to perform spill operation. */
+    private final ExecutorService ioExecutor =
+            Executors.newSingleThreadExecutor(
+                    new ThreadFactoryBuilder()
+                            .setNameFormat("ProducerMergePartitionFileWriter Spiller")
+                            .setUncaughtExceptionHandler(FatalExitExceptionHandler.INSTANCE)
+                            .build());
+
+    /** File channel to write data. */
+    private final FileChannel dataFileChannel;
+
+    /** Records the current writing location. */
+    private long totalBytesWritten;
+
+    private final PartitionFileIndex partitionFileIndex;
+
+    public ProducerMergePartitionFileWriter(
+            Path dataFilePath, PartitionFileIndex partitionFileIndex) {
+        LOG.info("Creating partition file " + dataFilePath);
+        try {
+            this.dataFileChannel =
+                    FileChannel.open(
+                            dataFilePath, StandardOpenOption.CREATE_NEW, StandardOpenOption.WRITE);
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to create file channel.", e);
+        }
+        this.partitionFileIndex = partitionFileIndex;
+    }
+
+    @Override
+    public CompletableFuture<Void> write(List<SubpartitionNettyPayload> toWriteBuffers) {
+        List<NettyPayload> buffersToSpill =
+                toWriteBuffers.stream()
+                        .map(SubpartitionNettyPayload::getSegmentNettyPayloads)
+                        .flatMap(
+                                (Function<List<SegmentNettyPayload>, Stream<SegmentNettyPayload>>)
+                                        Collection::stream)
+                        .map(SegmentNettyPayload::getNettyPayloads)
+                        .flatMap(
+                                (Function<List<NettyPayload>, Stream<NettyPayload>>)
+                                        Collection::stream)
+                        .collect(Collectors.toList());
```

Comment:
Then why do we need the complicated `SubpartitionNettyPayload` and `SegmentNettyPayload` at the first place?

Reply:
Because the subpartition id and segment id are used in the hash mode partition file. While the producer merge mode does not need that.